builtin/logical/pki/path_acme_account.go

```diff
@@ -230,18 +242,54 @@ func (b *backend) acmeNewAccountCreateHandler(acmeCtx *acmeContext, userCtx *jws
 		return formatAccountResponse(acmeCtx, accountByKey), nil
 	}
 
+	var eab *eabType
+	var eabData map[string]interface{}
+	if acmeCtx.requireEab {
+		eabDataRaw, ok := data["externalAccountBinding"]
+		if !ok {
+			return nil, fmt.Errorf("%w: externalAccountBinding missing from payload", ErrExternalAccountRequired)
+		}
+		eabData, ok = eabDataRaw.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("%w: externalAccountBinding field was unparseable", ErrMalformed)
+		}
+
+		eab, err = b.acmeState.verifyEabPayload(acmeCtx, userCtx, r.Path, eabData)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load account by thumbprint: %w", err)
+		}
+	}
+
 	// TODO: Limit this only when ToS are required or set by the operator, since we don't have a
 	//       ToS URL in the directory at the moment, we can not enforce this.
 	//if !termsOfServiceAgreed {
 	//	return nil, fmt.Errorf("terms of service not agreed to: %w", ErrUserActionRequired)
 	//}
 
-	accountByKid, err := b.acmeState.CreateAccount(acmeCtx, userCtx, contact, termsOfServiceAgreed)
+	if eab != nil {
+		// We delete the EAB to prevent future re-use after associating it with an account, worst
+		// case if we fail creating the account we simply nuked the EAB which they can create another
+		// and retry
+		wasDeleted, err := b.acmeState.DeleteEab(acmeCtx.sc, eab.KeyID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to delete eab reference: %w", err)
+		}
+
+		if !wasDeleted {
+			// Something consumed our EAB before we did bail...
+			return nil, fmt.Errorf("eab was already used: %w", ErrUnauthorized)
+		}
+	}
+
+	accountByKid, err := b.acmeState.CreateAccount(acmeCtx, userCtx, contact, termsOfServiceAgreed, eab)
 	if err != nil {
+		if eab != nil {
+			return nil, fmt.Errorf("failed to create account, EAB key that was used is no longer available: %w", err)
```

Comment:
I'd perhaps suggest stating:

```
failed to create account: %w; the EAB key used for this request has been deleted as a result of this operation; fetch a new EAB key before retrying
```

or similar? I think its unclear the implications of the middle part now (EAB key ... available) as it doesn't indicate the action the user should take on retry, and makes it sound like the error that follows is the reason why this key is no longer available.

Reply:
Updated in https://github.com/hashicorp/vault/commit/def8d7fdfdf66bcb63e5febee33089c3c6fcf3d0